### PR TITLE
ssl_peer_fingerprint_verification for self-signed certs

### PR DESCRIPTION
**Pull request by hh**
This pull request is mainly to discuss approaches to ssl fingerprint or certificate checking when self-signed cert Common Names do not match the FDQN or IP of the host we are connecting to.

### CODE REVIEW

File: lib/winrm/http/transport.rb [hunk 1]
```diff
@@ -47,6 +49,26 @@ def send_request(message)
...
+        # [23] pry(#<WinRM::HTTP::HttpSSL>)> winrm_fingerprint = OpenSSL::Digest::SHA1.new(resp.peer_cert.to_der).to_s.upcase
+        #  => "8014A7E5F7316F2E8D29866C5CF2CE348F2F5B67"
+        # Maybe we could check our cert chain for a self-signed cert matching here as well
+        # We should probably integrate this into a verify_self_signed_fingerprint!
+        # or something
+        if @ssl_peer_fingerprint
+          conn_fingerprint = OpenSSL::Digest::SHA1.new(
+            resp.peer_cert.to_der).to_s.upcase
+          if @ssl_peer_fingerprint != conn_fingerprint
+            @logger.fatal("ssl fingerprint mismatch!!!!\n")
```
Comment (hh) [inline lib/winrm/http/transport.rb]: This verifies connections ongoing.  See WinRb/WinRM#151

File: lib/winrm/http/transport.rb [hunk 2]
```diff
@@ -67,6 +89,28 @@ def no_ssl_peer_verification!
...
+        # Not sure if this is the right place to be making
+        # a probe / tcp connection to grab the cert and compare during
+        # initialize....
+        # We could just analyize the fingerprint after first request
+        noverify_peer_context = OpenSSL::SSL::SSLContext.new
+        noverify_peer_context.verify_mode = OpenSSL::SSL::VERIFY_NONE
+        tcp_connection = TCPSocket.new(@endpoint.host, @endpoint.port)
+        shady_ssl_connection = OpenSSL::SSL::SSLSocket.new(tcp_connection,
+                                                           noverify_peer_context)
+        shady_ssl_connection.connect
```
Comment (hh) [inline lib/winrm/http/transport.rb]: ssl_fingerprint_verification! will make an outbound tcp connection before sending any traffic to verify our cert fingerprint. See WinRb/WinRM#151

File: lib/winrm/http/transport.rb [hunk 3]
```diff
@@ -47,6 +49,26 @@ def send_request(message)
         resp = @httpcli.post(@endpoint, message, hdr)
```
Comment (hh) [inline lib/winrm/http/transport.rb]: At the time of this call to POST our first message, we don't know if the fingerprint will match, but we go ahead and send the http request.

We have already connected once in ssl_peer_fingerprint_verification! but if we had a complex malicious-in-the-middle scenario, they could wait for the initial verify then redirect all subsequent connections.

I'm not sure of all the initial http soap requests winrm makes, but I'd like to verify that we could fail after that first request and still be safe. It's obviously fine for negotiate and kerberos, but I'm unsure if the first request for plaintext would send anything.

File: lib/winrm/http/transport.rb [hunk 4]
```diff
@@ -67,6 +73,22 @@ def no_ssl_peer_verification!
...
+        noverify_peer_context.verify_mode = OpenSSL::SSL::VERIFY_NONE
+        tcp_connection = TCPSocket.new(@endpoint.host, @endpoint.port)
+        shady_ssl_connection = OpenSSL::SSL::SSLSocket.new(tcp_connection,
+                                                           noverify_peer_context)
+        shady_ssl_connection.connect
+        connection_cert = shady_ssl_connection.peer_cert_chain.last
+        conn_fingerprint=OpenSSL::Digest::SHA1.new(connection_cert.to_der).to_s.upcase
+        if @ssl_peer_fingerprint == conn_fingerprint
+          @logger.info("ssl fingerprint #{@ssl_peer_fingerprint} verified\n")
+          no_ssl_peer_verification!
```
Comment (hh) [inline lib/winrm/http/transport.rb]: Do note that we have to turn off ssl_peer_verification for this to work.

**Comment by mwrock** [comment]
After spending a good chunk of time writing [this winrm blog post](http://www.hurryupandwait.io/blog/understanding-and-troubleshooting-winrm-connection-and-authentication-a-thrill-seekers-guide-to-adventure) and most of the time getting some SSL scenarios working, I'm even more of a fan of this idea.

For instance, using `New-SelfSignedCertificate` available on 2012 and up does not work if exported to a PEM and then passed to the `:ca_trust_path` because it is not identified as a CA. This also includes the RDP cert already on the box. Its a pain to either download the tooling or supply the 50 lines of C# needed to add that constraint.

If we could simply pass a known thumbprint it would be much less friction than throwing self signed certs around and much safer than HTTP or disabled peer verification.

One could also submit a small vagrant pr to allow adding HTTPS attributes to a Vagrantfile making vagrant scenarios easier to secure and reducing the need to allow unencrypted basic auth.

We do need to add some tests and get the rubocop happy.

**Comment by sneal** [comment]
@mwrock What about 50 lines of PowerShell that just uses the .NET framework classes to do the work of generating the appropriate self signed cert? Everyone already needs a PS script to enable WinRM anyway.

**Comment by mwrock** [comment]
Yeah powershell works too :) Here is a [great powershell function](https://gallery.technet.microsoft.com/scriptcenter/Self-signed-certificate-5920a7c6) that has more granular control than `New-SelfSignedCertificate` which works great with an openssl client if you ad `-IsCA $true`.

However, its one more thing to download or pack into your image prep artifacts.

This model proposed by @hh is very similar to the way one connects to vSphere using the Fog API. With this, one could reuse the cert that windows creates for remote desktop. Just removes another layer of friction.

**Comment by hh** [comment]
@mwrock I cleaned up the rubocop, I'm not yet familiar enough with the testing framework. I'm taking a look now, but any useful/pointed directions would be great.

File: lib/winrm/http/transport.rb [hunk 5]
```diff
@@ -40,12 +40,13 @@ def initialize(endpoint)
...
+        ssl_peer_fingerprint_verification!
         log_soap_message(message)
-        hdr = {
-          'Content-Type' => 'application/soap+xml;charset=UTF-8',
-          'Content-Length' => message.length }
+        hdr = { 'Content-Type' => 'application/soap+xml;charset=UTF-8',
+                'Content-Length' => message.length }
         resp = @httpcli.post(@endpoint, message, hdr)
         log_soap_message(resp.http_body.content)
+        verify_ssl_fingerprint(resp.peer_cert)
```
Comment (mwrock) [inline lib/winrm/http/transport.rb]: You need a `if @ssl_peer_fingerprint` here so it is not called for normal requests. Or perhaps even better, return from that method if its `nil`.

File: lib/winrm/http/transport.rb [hunk 6]
```diff
@@ -67,6 +68,34 @@ def no_ssl_peer_verification!
         @httpcli.ssl_config.verify_mode = OpenSSL::SSL::VERIFY_NONE
       end
 
+      # SSL Peer Fingerprint Verification prior to connecting
+      def ssl_peer_fingerprint_verification!
+        return unless @ssl_peer_fingerprint && ! @ssl_peer_fingerprint_verified
+        connection_cert = shady_ssl_connection.peer_cert_chain.last
```
Comment (mwrock) [inline lib/winrm/http/transport.rb]: should `sysclose` be called on the SSL socket when done?

Comment (hh) [inline lib/winrm/http/transport.rb]: Wouldn't hurt.

**Comment by mwrock** [comment]
Merged in with #170

**Pull request closed without merging**